DOCS 640: Update ML Service and Vision Service (#1074)
viamrobotics/docs: docs/services/vision/detection.md

# SMEs: Bijan, Khari
---

> _Changed in [RDK v0.3.0 and API v0.2.0](/appendix/release-notes/#2-may-2023)_

Author: ```suggestion
_Changed in [RDK v0.3.0 and API v0.1.118](/appendix/release-notes/#2-may-2023)_
```

- `confidence` (float): specifies the confidence of the assigned label.
  Between `0.0` and `1.0`, inclusive.

You can use the following types of detectors:

- [**color_detector**](#configure-a-color_detector): A heuristic detector that draws boxes around objects according to their hue (does not detect black, gray, and white).

> - [**mlmodel**](#configure-a-mlmodel-detector): A machine learning detector that draws bounding boxes according to the specified .tflite model file available on the robot’s hard drive.

Reviewer: Rather than being specifically about a tflite model, `mlmodel` should be able to use any ML Model service, whether it's a tflite model, or another model that the ML Model Service will one day support.

> - [**mlmodel**](#configure-a-mlmodel-detector): A machine learning detector that draws bounding boxes according to the specified .tflite model file available on the robot’s hard drive.

Author: ```suggestion
- [**mlmodel**](#configure-a-mlmodel-detector): A machine learning detector that draws bounding boxes according to a ML model available on the robot’s hard drive.
```

> - [**tflite_detector**](#configure-a-tflite_detector): A machine learning detector that draws bounding boxes according to the specified .tflite model file available on the robot’s hard drive.

Author: ```suggestion
- [**tflite_detector**](#configure-a-tflite_cpu-detector): A machine learning detector that draws bounding boxes according to the specified .tflite model file available on the robot’s hard drive.
```

> - [**tflite_detector**](#configure-a-tflite_cpu-detector): A machine learning detector that draws bounding boxes according to the specified .tflite model file available on the robot’s hard drive.

Author: ```suggestion
- [**tflite_cpu**](#configure-a-tflite_cpu-detector): A machine learning detector that draws bounding boxes according to the specified .tflite model file available on the robot’s hard drive.
```

If the color is not reliably detected, increase the `hue_tolerance_pct`.
{{< /alert >}}

{{< tabs >}}
{{% tab name="Builder" %}}

Navigate to the [robot page on the Viam app](https://app.viam.com/robots).
Click on the robot you wish to add the Vision Service to.
Select the **config** tab, and click on **Services**.

Scroll to the **Create Service** section.
To create a [Vision Service](/services/vision/):

1. Select `vision` as the **Type**.
2. Enter a name as the **Name**.

> 3. Select `ml_model` as the **Model**.

Reviewer: this should be `color_detector`

{{% /alert %}}

Click **Save config** and head to the **Components** tab.
Proceed to [Add a camera component and a "transform" model](#add-a-camera-component-and-a-transform-model).

## Configure a `tflite_detector`

A machine learning detector that draws bounding boxes according to the specified tensorflow-lite model file available on the robot’s hard drive.

### Create the ML Model Service

> Navigate to the [robot page on the Viam app](https://app.viam.com/robots), then create an ML Model Service for the classifier model:

Reviewer: for the detector model

{{< tabs >}}
{{% tab name="Builder" %}}

> Click on the robot you wish to add the classifier to.

Reviewer: to add the detector to

Select the **config** tab, and click on **Services**.

Scroll to the **Create Service** section:

> 1. Select `ml_model` as the **Type**.

Reviewer: the type is just "mlmodel" rather than "ml_model"

It's a bit confusing, because the model in the vision service is called "ml_model" ...

Reviewer: fixed in the code so everything is "mlmodel"

2. Enter a name as the **Name**.
3. Select `tflite_cpu` as the **Model**.
4. Click **Create Service**.

In your ML Model Service's panel, fill in the **Attributes** field.

``` json {class="line-numbers linkable-line-numbers"}
{
      "model_path": "${packages.<model-name>}/<model-name>.tflite",
      "label_path": "${packages.<model-name>}/labels.txt",
      "num_threads": <number>
}
```

{{% /tab %}}
{{% tab name="JSON Template" %}}

Add the classifier ML model object to the services array in your raw JSON configuration:

``` json {class="line-numbers linkable-line-numbers"}
"services": [
  {

>     "name": "<classifier_name>",

Reviewer: detector_name

    "type": "ml_model",
    "model": "tflite_cpu",
    "attributes": {
      "model_path": "${packages.<model-name>}/<model-name>.tflite",
      "label_path": "/${packages.<model-name>}/labels.txt",
      "num_threads": <number>
    }
  },
  ... // Other services
]
```

{{% /tab %}}
{{% tab name="JSON Example" %}}

```json {class="line-numbers linkable-line-numbers"}
"services": [
  {
    "name": "person_detector",

>     "type": "ml_model",

Reviewer: type should be mlmodel

{{% /alert %}}

Click **Save config** and head to the **Components** tab.
Proceed to [Add a camera component and a "transform" model](#add-a-camera-component-and-a-transform-model).

## Configure a `mlmodel` detector

A machine learning detector that draws bounding boxes according to the specified tensorflow-lite model file available on the robot’s hard drive.
To create a `mlmodel` classifier, you need an [ML Model Service with a suitable model](../../ml).

Navigate to the [robot page on the Viam app](https://app.viam.com/robots).
Click on the robot you wish to add the Vision Service to.
Select the **config** tab, and click on **Services**.

Scroll to the **Create Service** section.

{{< tabs >}}
{{% tab name="Builder" %}}

1. Select `vision` as the **Type**.
2. Enter a name as the **Name**.
3. Select `mlmodel` as the **Model**.
4. Click **Create Service**.

![Create Vision Service for mlmodel](../img/mlmodel.png)

In your Vision Service's panel, fill in the **Attributes** field.

``` json {class="line-numbers linkable-line-numbers"}
{

>   "detector_name": "<classifier_name>"

Reviewer: "mlmodel_name": "<detector_name>"

>   "detector_name": "<classifier_name>"

Author: ```suggestion
  "mlmodel_name": "<detector_name>"
```

These requirements are satisfied by a few publicly available model architectures including EfficientDet, MobileNet, and SSD MobileNet V1.
You can use one of these architectures or build your own.

### Create the Vision Service

Create another service:

{{< tabs >}}
{{% tab name="Builder" %}}

1. Select `vision` as the **Type**.
2. Enter a name as the **Name**.
3. Select `ml_model` as the **Model**.
4. Click **Create Service**.

In your Vision Service's panel, fill in the **Attributes** field.

``` json {class="line-numbers linkable-line-numbers"}
{

>   "ml_model_name": "<classifier_name>"

Reviewer: detector_name

}
```

{{% /tab %}}
{{% tab name="JSON Template" %}}

Add the Vision Service object to the services array in your raw JSON configuration:

``` json {class="line-numbers linkable-line-numbers"}
"services": [
  {
    "name": "<service_name>",
    "type": "vision",
    "model": "ml_model",
    "attributes": {

>       "ml_model_name": "<classifier_name>"

Reviewer: detector_name
